lib/src/protocol_instance.rs

```diff
@@ -114,6 +115,42 @@ pub fn assemble_protocol_instance(
         TxHash::from(keccak(input.taiko.tx_data.as_slice()))
     };
 
+    // If the passed in chain spec contains a known chain id, the chain spec NEEDS to match the
+    // one we expect, because the prover could otherwise just fill in any values.
+    // The chain id is used because that is the value that is put onchain,
+    // and so all other chain data needs to be derived from it.
+    // For unknown chain ids we just skip this check so that tests using test data can still pass.
+    // TODO: we should probably split things up in critical and non-critical parts
+    // in the chain spec itself so we don't have to manually all the ones we have to care about.
+    if let Some(verified_chain_spec) =
+        SupportedChainSpecs::default().get_chain_spec_with_chain_id(input.chain_spec.chain_id)
```

Comment:
Seems it also relies on fixed image_id checking on-chain (in sgx it's mrenclave, in r0/sp1 it's image_id), as the default value can be override by user.

Reply:
Yep! But the config is injected in the code, so it's part of the program itself for both SGX and zkVMs so it will already be part of whatever system is used to ensure the correct code is executed.